### ISSUE TITLE
[FEA] dataframe.mode() axis parameter not supported

### ISSUE THREAD

**Issue by dawilliams-nvidia** [rapidsai/cudf#11925]
**Is your feature request related to a problem? Please describe.**
The cudf documentation lists the mode function's {axis=0,1} parameter, but then has an additional note that "axis parameter is currently not supported.":  https://docs.rapids.ai/api/cudf/stable/api_docs/api/cudf.DataFrame.mode.html 

**Describe the solution you'd like**
Is this feature possible to be implemented? Or is there an underlying challenge with replicating this pandas functionality in cudf?

**Additional context**
NVIDIA Solutions Architect, filing on behalf of customer

**Comment by brandon-b-miller** [comment]
It seems like it would be possible to write a numba kernel that consumes a matrix-like data structure (possibly generated from the input dataframe) and computes the mode across the rows, however I think the challenge for us in cuDF would be making sure that kernel (or family of kernels) matches pandas for all the edge cases involving dataframes with mixed dtypes, nulls, nans, and the like. We'd also want to think about the UX of dispatching to cupy for some row-wise operations and a custom numba kernel which requires compilation for others.

**Comment by beckernick** [comment]
I agree. This is probably in scope for CuPy, as SciPy explicitly implements it (with a fairly interesting return type). Given this operation is not columnar in nature, I think it makes sense to implement this in the Python layer by dispatching to a potential future CuPy implementation.

**Comment by beckernick** [comment]
Per offline discussion with @dawilliams-nvidia , limited support for only numeric types would be sufficient

**Comment by bdice** [comment]
If numeric-only types are sufficient here, I would suggest the same thing as I mentioned here: https://github.com/rapidsai/cudf/issues/11570#issuecomment-1221119749

We can implement something like `axis=1` by returning `df.T.mode(axis=0).T`. The explicit transposes will only work for dataframes with similar dtypes (I am not exactly sure what qualifies for like-casting on tranpose) but it would permit a reasonable subset of `axis=1` use cases to be performed in a fairly efficient manner. The two transposes are not ideal for speed, but neither is row-wise operation. This might be an okay compromise between GPU perf and usability.

**Comment by beckernick** [comment]
We changed transpose to return a contiguous buffer, right? I think that allocating thousands of columns will be a non-starter.

**Comment by shwina** [comment]
> We changed transpose to return a contiguous buffer, right?

Indeed we did.

**Comment by bdice** [comment]
I see. An alternative would be to allocate a single column that is matrix-like (has `columns*rows` entries), copy all the entries there (no need to transpose), and add an auxiliary column of row labels. Then do a groupby count over both columns `(row_label, value)` to get the counts of each value in each row. Take the result and do a groupby argmax over `row_label` to get the index of the mode of each row, and gather the corresponding values to return.
